Tests/Migration/DataPipelineMigrationAssistantTests.swift

```diff
@@ -150,4 +139,319 @@ class DataPipelineMigrationAssistantTests: UnitTest {
+
+        XCTAssertNotNil(migrationAssistant.getAndProcessTask(for: givenCreatedTask, siteId: testSiteId))
+        XCTAssertEqual(backgroundQueueMock.deleteProcessedTaskCallsCount, 0)
+
+        // When device attribute is invalid
+        let pushTokenTaskDataWithInvalidJsonString = RegisterPushNotificationQueueTaskData(profileIdentifier: String.random, attributesJsonString: "{\"foo\": \"bar\"}")
+
+        guard let jsonData = try? JSONEncoder().encode(pushTokenTaskDataWithInvalidJsonString) else {
+            XCTFail("Failed to create task data")
+            return
+        }
+        backgroundQueueMock.getTaskDetailReturnValue = TaskDetail(data: jsonData, taskType: givenType, timestamp: dateUtilStub.now)
+
+        XCTAssertNotNil(migrationAssistant.getAndProcessTask(for: givenCreatedTask, siteId: testSiteId))
+        XCTAssertEqual(backgroundQueueMock.deleteProcessedTaskCallsCount, 0)
+    }
+
+    func test_givenInvalidDeletePushToken_expectTaskNotProcessed() {
+        let givenType = QueueTaskType.deletePushToken
+
+        guard let givenCreatedTask = createTaskAndStoreInInventory(forType: givenType) else {
+            XCTFail("Failed to create task")
+            return
+        }
+
+        backgroundQueueMock.getTaskDetailReturnValue = TaskDetail(data: Data(), taskType: givenType, timestamp: dateUtilStub.now)
+
+        XCTAssertNotNil(migrationAssistant.getAndProcessTask(for: givenCreatedTask, siteId: testSiteId))
+        XCTAssertEqual(backgroundQueueMock.deleteProcessedTaskCallsCount, 0)
+    }
+
+    func test_InvalidTrackPushMetric_expectTaskNotProcessed() {
+        let givenType = QueueTaskType.trackPushMetric
+
+        guard let givenCreatedTask = createTaskAndStoreInInventory(forType: givenType) else {
+            XCTFail("Failed to create task")
+            return
+        }
+        backgroundQueueMock.getTaskDetailReturnValue = TaskDetail(data: Data(), taskType: givenType, timestamp: dateUtilStub.now)
+
+        XCTAssertNotNil(migrationAssistant.getAndProcessTask(for: givenCreatedTask, siteId: testSiteId))
+        XCTAssertEqual(backgroundQueueMock.deleteProcessedTaskCallsCount, 0)
+    }
+}
+
+extension DataPipelineMigrationAssistantTests {
+    private func createTaskAndStoreInInventory(forType type: QueueTaskType) -> QueueTaskMetadata? {
+        guard let fileManagerQueueStorage = queueStorage as? FileManagerQueueStorage else {
+            XCTFail("queueStorage could not be cast to FileManagerQueueStorage")
+            return nil
+        }
+
+        guard let givenCreatedTask = fileManagerQueueStorage.create(siteId: testSiteId, type: type.rawValue, data: Data(), groupStart: nil, blockingGroups: nil).createdTask else {
+            XCTFail("Failed to create task")
+            return nil
+        }
+        var inventory: [QueueTaskMetadata] = []
+        inventory.append(givenCreatedTask)
+        backgroundQueueMock.getAllStoredTasksReturnValue = inventory
+
+        return givenCreatedTask
+    }
```

Comment:
Could you explain more about how this function works? 

Because the BQ is mocked, I don't think that any data is being read from the file system. I think that we are saying files in the file system with the intention of the BQ reading those tasks from the file system to process? 

Could we not mock the BQ? I think that our migration tests could be higher quality if we are running the BQ implementation rather then a mock.

Reply:
You're right! I have fixed the implementation. Thanks for calling out.